modules/apps/29-fee/keeper/escrow_test.go

```diff
@@ -0,0 +1,236 @@
+			} else {
+				suite.Require().Error(err)
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestPayFee() {
+	var (
+		err            error
+		ackFee         *sdk.Coin
+		receiveFee     *sdk.Coin
+		timeoutFee     *sdk.Coin
+		packetId       *channeltypes.PacketId
+		reverseRelayer sdk.AccAddress
+		forwardRelayer sdk.AccAddress
+		refundAcc      sdk.AccAddress
+	)
+
+	// refundAcc does not have balance for the following Coin
+	validChannelId := "channel-0"
+	validSeq := uint64(1)
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expPass  bool
+	}{
+		{
+			"success", func() {}, true,
+		},
+		{
+			"fee not found for packet", func() {
+				// setting packetId with an invalid sequence of 2
+				packetId = &channeltypes.PacketId{ChannelId: validChannelId, PortId: types.PortKey, Sequence: uint64(2)}
+			}, false,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		suite.Run(tc.name, func() {
+			suite.SetupTest() // reset
+
+			// setup
+			refundAcc = suite.chainA.SenderAccount.GetAddress()
+			reverseRelayer = suite.chainA.SenderAccount.GetAddress()
+			forwardRelayer = suite.chainA.SenderAccount.GetAddress()
+
+			validCoin := &sdk.Coin{Denom: sdk.DefaultBondDenom, Amount: sdk.NewInt(100)}
+			ackFee = validCoin
+			receiveFee = validCoin
+			timeoutFee = validCoin
+			packetId = &channeltypes.PacketId{ChannelId: validChannelId, PortId: types.PortKey, Sequence: validSeq}
+			fee := &types.Fee{ackFee, receiveFee, timeoutFee}
+
+			// escrow the packet fee & store the fee in state
+			identifiedPacketFee := types.IdentifiedPacketFee{PacketId: packetId, Fee: fee, Relayers: []string{}}
+
+			err = suite.chainA.GetSimApp().IBCFeeKeeper.EscrowPacketFee(suite.chainA.GetContext(), refundAcc, identifiedPacketFee)
+			suite.Require().NoError(err)
+
+			tc.malleate()
+
+			err = suite.chainA.GetSimApp().IBCFeeKeeper.DistributeFee(suite.chainA.GetContext(), refundAcc, forwardRelayer, reverseRelayer, packetId)
+
+			if tc.expPass {
+				suite.Require().NoError(err)
+				hasFeeInEscrow := suite.chainA.GetSimApp().IBCFeeKeeper.HasFeeInEscrow(suite.chainA.GetContext(), packetId.ChannelId, packetId.Sequence)
+				// there should no longer be a fee in escrow for this packet
+				suite.Require().False(hasFeeInEscrow)
+			} else {
+				suite.Require().Error(err)
+				hasFeeInEscrow := suite.chainA.GetSimApp().IBCFeeKeeper.HasFeeInEscrow(suite.chainA.GetContext(), packetId.ChannelId, 1)
+				// there should still be a fee in escrow for this packet
+				suite.Require().True(hasFeeInEscrow)
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestPayTimeoutFee() {
+	var (
+		err            error
+		ackFee         *sdk.Coin
+		receiveFee     *sdk.Coin
+		timeoutFee     *sdk.Coin
+		packetId       *channeltypes.PacketId
+		reverseRelayer sdk.AccAddress
+		refundAcc      sdk.AccAddress
+	)
+
+	// refundAcc does not have balance for the following Coin
+	validChannelId := "channel-0"
+	validSeq := uint64(1)
+
+	testCases := []struct {
```

Comment:
Test cases where `RecvFee+AckFee > TimoutFee` and vice versa after implementing capital efficient escrow